crates/networking/p2p/peer_handler.rs

```diff
@@ -131,59 +131,137 @@ impl PeerHandler {
             .flatten()
             .and_then(|headers| (!headers.is_empty()).then_some(headers))
             {
-                return Some(block_headers);
+                let block_hashes = block_headers
+                    .iter()
+                    .map(|header| header.compute_block_hash())
+                    .collect::<Vec<_>>();
+
+                if are_block_headers_chained(&block_headers, &block_hashes) {
+                    return Some((block_headers, block_hashes));
+                } else {
+                    warn!("Received invalid headers from peer, discarding peer {peer_id} and retrying...");
+                    self.remove_peer(peer_id).await;
+                }
             }
         }
         None
     }
 
-    /// Requests block bodies from any suitable peer given their block hashes
+    /// Internal method to request block bodies from any suitable peer given their block hashes
     /// Returns the block bodies or None if:
     /// - There are no available peers (the node just started up or was rejected by all other nodes)
     /// - No peer returned a valid response in the given time and retry limits
-    pub async fn request_block_bodies(&self, block_hashes: Vec<H256>) -> Option<Vec<BlockBody>> {
```

Comment:
The method's doc still reflects old method (it mentions retry limits) It should mention that it only performs one request from one peer